changelog.d/9522.bugfix

```diff
@@ -0,0 +1 @@
+Fix #9507 by hardening ResponseCache against errors in execution of to-be-cached functions.
```

Comment:
I think this should be the same message as #9358:

> Added a fix that invalidates cache for empty timed-out sync responses.

Please also rename `changelog.d/9358.misc` -> `changelog.d/9358.bugfix` per https://github.com/matrix-org/synapse/pull/9358#discussion_r585472307

Reply:
Is it possible to do that without freaking out towncrier?

Reply:
towncrier is not a bot, it is just a tool we run during release. So yes, we do this all the time.